upcloud/managed_database.go

```diff
@@ -0,0 +1,693 @@
+}
+
+// ManagedDatabaseComponent represents an accessible component within a service. The usage varies between service types
+type ManagedDatabaseComponent struct {
+	Component string `json:"component"`
+	Host      string `json:"host"`
+	Port      int    `json:"port"`
+	// Route describes how the component can be reached.
+	// 	public  = The component can be reached over public internet
+	// 	dynamic = The component can be only reached over utility network
+	Route string `json:"route"`
+	// Usage describes the role of the component.
+	// 	primary = The component is a primary (writable) instance in a cluster
+	// r	eplica = The component is a standby (read-only) instance in a cluster
+	Usage string `json:"usage"`
+}
+
+// ManagedDatabaseConnection represents a connection to the managed database instance. The available fields
+// vary between database types.
+type ManagedDatabaseConnection struct {
+	ApplicationName string        `json:"application_name"`
+	BackendStart    time.Time     `json:"backend_start"`
+	BackendType     string        `json:"backend_type"`
+	BackendXid      string        `json:"backend_xid"`
+	BackendXmin     string        `json:"backend_xmin"`
+	ClientAddr      string        `json:"client_addr"`
+	ClientHostname  string        `json:"client_hostname"`
+	ClientPort      int           `json:"client_port"`
+	DatId           int           `json:"dat_id"`
+	DatName         string        `json:"dat_name"`
+	Pid             int           `json:"pid"`
+	Query           string        `json:"query"`
+	QueryDuration   time.Duration `json:"query_duration"`
+	QueryStart      time.Time     `json:"query_start"`
+	State           string        `json:"state"`
+	StateChange     time.Time     `json:"state_change"`
+	Username        string        `json:"usename"`
+	UseSysId        int           `json:"usesysid"`
+	WaitEvent       string        `json:"wait_event"`
+	WaintEventType  string        `json:"waint_event_type"`
+	XactStart       time.Time     `json:"xact_start"`
+}
+
+// MarshalJSON implements json.Marshaler
+func (m *ManagedDatabaseConnection) MarshalJSON() ([]byte, error) {
+	type alias ManagedDatabaseConnection
+	out := struct {
+		alias
+		QueryDuration float64 `json:"query_duration"`
+	}{}
+	out.alias = alias(*m)
+	out.QueryDuration = m.QueryDuration.Seconds()
+	return json.Marshal(&out)
+}
+
+// UnmarshalJSON implements json.Unmarshaler
+func (m *ManagedDatabaseConnection) UnmarshalJSON(d []byte) error {
+	type alias ManagedDatabaseConnection
+	in := struct {
+		alias
+		QueryDuration float64 `json:"query_duration"`
+	}{alias: alias(*m)}
+	if err := json.Unmarshal(d, &in); err != nil {
+		return err
+	}
+	*m = ManagedDatabaseConnection(in.alias)
+	m.QueryDuration = time.Duration(in.QueryDuration * float64(time.Second))
+	return nil
+}
+
+// ManagedDatabaseMaintenance represents the set time of week when automatic maintenance operations are allowed
+type ManagedDatabaseMaintenance struct {
```

Comment:
Could make the relation to time explicit in the name of the type (`ManagedDatabaseMaintenanceTime`?), but not a big deal.

Reply:
yes, will rename

Reply:
implemented